Sources/WebService/AuthenticatedWebService/AuthenticatedWebService.swift

```diff
@@ -56,84 +56,83 @@ open class AuthenticatedWebService: WebService {
     self.configuration = configuration
     super.init(urlSession: urlSession)
   }
-  
+
   override public func execute<T>(urlRequest: URLRequest) -> AnyPublisher<T, Error> where T : Decodable {
     var urlRequest = urlRequest
-    var currentAccessToken: String?
 
-    authenticationQueue.sync {
-      currentAccessToken = self.tokenProvider.accessToken.value
+    func appendTokenAndExecute(accessToken: AccessToken) -> AnyPublisher<T, Error> {
+      urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
+      return super.execute(urlRequest: urlRequest)
+        .subscribe(on: DispatchQueue.global())
+        .eraseToAnyPublisher()
+    }
+
+    return Deferred {
+      self.tokenProvider.accessToken
+        .compactMap { $0 }
+        .setFailureType(to: Error.self)
+        .flatMap { accessToken -> AnyPublisher<T, Error> in
+          return appendTokenAndExecute(accessToken: accessToken)
+      }
     }
-    
-    guard let accessToken = currentAccessToken else {
-      return Fail<T, Error>(error: NetworkError.unauthorized).eraseToAnyPublisher()
+    .timeout(10, scheduler: DispatchQueue.main, customError: { NetworkError.timeout })
+    .catch { [weak self] error -> AnyPublisher<T, Error> in
+      guard let self = self else {
+        return Fail<T, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
+      }
+
+      if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
+        return self.retrySynchronizedTokenRefresh()
+          .flatMap {
+            appendTokenAndExecute(accessToken: $0)
+        }.eraseToAnyPublisher()
+      }
+      return Fail<T, Error>(error: error).eraseToAnyPublisher()
     }
-    
-    urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
-    
-    return super.execute(urlRequest: urlRequest)
-      .catch { [weak self] error -> AnyPublisher<T, Error> in
-        guard let self = self else {
-          return Fail<T, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
-        }
-
-        if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
-          self.retrySynchronizedTokenRefresh()
-
-          return self.execute(urlRequest: urlRequest)
-            .delay(for: 0.2, scheduler: self.authenticationQueue)
-            .eraseToAnyPublisher()
-        }
-        return Fail<T, Error>(error: error).eraseToAnyPublisher()
-      }.eraseToAnyPublisher()
+    .receive(on: DispatchQueue.main)
```

Comment:
same here

Reply:
It's a design decision that I took while creating the framework. For the consumer it doesn't change anything actually but I wanted fusion to provide a hassle-free development for the consumer. Therefore all the executions happen on the background thread and the deliveries happen on the main queue.

If at any point a consumer wants to use a different thread, they are free to use `subscribe or receive`